ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java

```diff
@@ -1530,6 +1532,102 @@ public void testReplicateBlobFailureDueToNotEnoughQuorum() throws Exception {
+        Assert.fail("Should return Ambry unavailable.");
+      } catch (ExecutionException e) {
+        Throwable t = e.getCause();
+        assertTrue("Cause should be RouterException", t instanceof RouterException);
+        assertEquals("ErrorCode mismatch", RouterErrorCode.AmbryUnavailable, ((RouterException) t).getErrorCode());
+      }
+
+      layout.getMockServers().forEach(mockServer -> mockServer.setServerErrorForAllRequests(null));
+
+      // Replicate Blob from the local replica to some remote replicas.
+      // choose one local replica as the source DataNode. Set the other two replicas as Replica_Unavailable
+      DataNodeId sourceDataNode = null;
+      for (MockServer server : layout.getMockServers()) {
+        if (server.getDataCenter().equals(localDcName)) {
+          if (sourceDataNode == null) {
+            sourceDataNode = mockClusterMap.getDataNodeId(server.getHostName(), server.getHostPort());
+          } else {
+            server.setServerErrorForAllRequests(ServerErrorCode.Replica_Unavailable);
+          }
+        }
+      }
+      // since there is only one local available, router will replicate the blob to at least 1 remote replicas.
+      // Replicating the four chunks. Also replicate the composite metadata blob.
+      GetBlobResult localGetChunkResult = router.getBlob(compositeBlobId,
+          new GetBlobOptionsBuilder().operationType(GetBlobOptions.OperationType.BlobChunkIds).build(), null, null)
+          .get();
+      List<StoreKey> chunkIds = localGetChunkResult.getBlobChunkIds();
+      for (StoreKey chunk : chunkIds) {
+        router.replicateBlob(chunk.getID(), serviceId, sourceDataNode).get();
+      }
+      router.replicateBlob(compositeBlobId, serviceId, sourceDataNode).get();
+
+      // set unavailable to all local replicas. So router.getBlob will get the Blob from remote replicas.
+      for (MockServer server : layout.getMockServers()) {
+        if (server.getDataCenter().equals(localDcName)) {
+          server.setServerErrorForAllRequests(ServerErrorCode.Replica_Unavailable);
+        }
+      }
+      GetBlobResult remoteGetBlobResult = router.getBlob(compositeBlobId, new GetBlobOptionsBuilder().build(), null, null).get();
+
+      // verify BlobProperties
+      Assert.assertEquals(remoteGetBlobResult.getBlobInfo().getBlobProperties(), localProperties);
+      // verify userMetadata
+      Assert.assertTrue(Arrays.equals(remoteGetBlobResult.getBlobInfo().getUserMetadata(), putUserMetadata));
+      // verify data content
+      RetainingAsyncWritableChannel retainingAsyncWritableChannel = new RetainingAsyncWritableChannel();
+      remoteGetBlobResult.getBlobDataChannel().readInto(retainingAsyncWritableChannel, null).get();
+      InputStream input = retainingAsyncWritableChannel.consumeContentAsInputStream();
+      retainingAsyncWritableChannel.close();
+      byte[] content = Utils.readBytesFromStream(input, (int) PUT_CONTENT_SIZE);
```

Comment:
There's no need to cast since it's int already right?

Reply:
You are right.